src/app/query/nonce/account-nonces.utils.ts

```diff
@@ -0,0 +1,56 @@
+import toast from 'react-hot-toast';
+import { AddressNonces } from '@stacks/blockchain-api-client/lib/generated';
+import { MempoolTransaction, Transaction } from '@stacks/stacks-blockchain-api-types';
+
+export function getNextNonce(
+  accountNonces: AddressNonces,
+  confirmedTransactions: Transaction[],
+  pendingTransactions: MempoolTransaction[]
+) {
+  const lastExecutedNonce = accountNonces.last_executed_tx_nonce;
+  const lastMempoolTxNonce = accountNonces.last_mempool_tx_nonce;
+  const possibleNextNonce = accountNonces.possible_next_nonce;
+
+  const detectedMissingNonces = accountNonces.detected_missing_nonces;
+  const hasMissingNonces = detectedMissingNonces?.length > 0;
+  const firstMissingNonce = detectedMissingNonces?.sort()[0];
+
+  const hasPendingTxs = pendingTransactions?.length > 0;
+  const pendingNonces = pendingTransactions.map(tx => tx.nonce);
+  const lastPendingTxNonce = pendingTransactions[0]?.nonce;
+
+  const pendingNoncesIncludesPossibleNextNonceFromApi = pendingNonces.includes(possibleNextNonce);
+
+  // Validate a 0 nonce returned from the api
+  if (possibleNextNonce === 0 && confirmedTransactions.length > 0)
+    return confirmedTransactions[0]?.nonce + 1;
+
+  // Compare missing nonces with pending tx nonces bc the api
+  // is often slow to update the missing nonces array
+  if (hasMissingNonces && hasPendingTxs) {
+    const remainingMissingNonces = detectedMissingNonces.filter(
+      nonce => !pendingNonces.includes(nonce)
+    );
+    return remainingMissingNonces.length > 0
+      ? Math.min(...remainingMissingNonces)
+      : pendingNoncesIncludesPossibleNextNonceFromApi
+      ? lastPendingTxNonce + 1
+      : possibleNextNonce;
+  }
+
+  // If there is a missing nonce, but no pending txs (and as a validation check, it is not
+  // equal to the last executed nonce), just return the first missing nonce
+  if (hasMissingNonces && !hasPendingTxs && lastExecutedNonce !== firstMissingNonce)
+    return firstMissingNonce;
+
+  // It is possible that the detected_missing_nonces array won't update fast
+  // enough and we will continue here w/out knowing there is a missing nonce.
+  // This is a backup check that might help...
+  if (!!(lastExecutedNonce && lastMempoolTxNonce) && lastMempoolTxNonce - lastExecutedNonce > 1) {
+    toast.error('Possible missing nonce detected');
```

Comment:
Not sure if we should show a toast here? Just a safety check but not totally sure how it will perform. I will do some testing on it today.

Reply:
Yeah this is a side-effect I'd keep out of the calculation method.

I wonder if this fn could return
```ts
{ none: number, calcStyle: 'api-suggestion' | 'local-fallback' | 'etc' }
```

So that the invoker knows through what means we landed on the resulting nonce (and can throw a toast accordingly).

Reply:
Interesting idea, let me give it a shot.